File: retrocpu_boot_monitor/mn1613/test/handshake/handshake_read_memory_test.ts
/**
 * g_hshk_read_memory（IO→CPU コマンド 13h）
 * 根拠: HandShake.mdc「メモリ読み出し」/ boot_monitor.mdc / test_framework.mdc
 */
import {
  createSessionFromSettings,
  expect,
  test,
  type IoBoardHandshakeMock,
  type Mn1613AsmSession,
} from "../../../../retrocpu_test_framework/src/index.js";
import {
  mn1613MonHandshakeSettings,
  withMn1613CpuLog,
} from "../mn1613_mon_settings.js";

const BASE_REGS = {
  R2: 0x2222,
  R3: 0x3333,
  R4: 0x4444,
} as const;

/** ユーザ RAM 先頭（ワード）。バイトアドレス = 0x3000 */
const WORD_ADDR = 0x1800;
const BYTE_ADDR = WORD_ADDR * 2;

const session: Mn1613AsmSession = createSessionFromSettings(
  withMn1613CpuLog(
    {
      ...mn1613MonHandshakeSettings,
      maxCycles: 250_000_000,
      ioMock: [{ type: "handshake", timeoutMs: 30_000, syncIrq2: false }],
    },
    import.meta.url,
  ),
);

/** 13h メモリ転送テストの上限（MN1613 前提で 8KB 以内） */
const MAX_MEMREAD_TEST_BYTES = 0x2000;
/** 実行時間を抑えるための実テストサイズ（512B） */
const MEMREAD_TEST_BYTES = 0x0200;

/**
 * 13h ヘッダ（cmd + addr32 BE + count32 BE + パッド）。
 * @param byteAddr バイトアドレス
 * @param byteCount バイト数
 * @returns IO→CPU 先頭フレーム
 */
function memReadHeader(byteAddr: number, byteCount: number): Uint8Array {
  return Uint8Array.from([
    0x13,
    (byteAddr >>> 24) & 0xff,
    (byteAddr >>> 16) & 0xff,
    (byteAddr >>> 8) & 0xff,
    byteAddr & 0xff,
    (byteCount >>> 24) & 0xff,
    (byteCount >>> 16) & 0xff,
    (byteCount >>> 8) & 0xff,
    byteCount & 0xff,
    0,
  ]);
}

/**
 * g_main 済み＋ ioMock handshake で 1 ケースを実行する。
 * @param fn 本体
 */
async function withCase(
  fn: (s: Mn1613AsmSession, mock: IoBoardHandshakeMock) => Promise<void>,
): Promise<void> {
  session.reload();
  try {
    await session.runInit();
    await fn(session, session.requireHandshakeMock());
  } finally {
    await session.detachIoMock();
  }
}

/**
 * IO が HSHK_IN_REQ を上げるまで待つ。
 * @param mock IO モック
 * @param timeoutMs 上限 ms
 */
async function waitReq1(
  mock: IoBoardHandshakeMock,
  timeoutMs = 2000,
): Promise<void> {
  const t0 = Date.now();
  while (mock.bus.HSHK_IN_REQ !== 1) {
    if (Date.now() - t0 > timeoutMs) {
      throw new Error("timeout waiting HSHK_IN_REQ");
    }
    await new Promise<void>((resolve) => setTimeout(resolve, 0));
  }
}

/**
 * 13h を IRQ ハンドラ経由で実行する。
 * @param mock IO モック
 * @param byteAddr 開始バイトアドレス
 * @param byteCount CPU→IO で受け取るデータ長
 * @returns CPU→IO 応答
 */
async function callRead(
  mock: IoBoardHandshakeMock,
  byteAddr: number,
  byteCount: number,
): Promise<Uint8Array> {
  const io = mock.exchangeWithCpu(
    memReadHeader(byteAddr, byteCount),
    byteCount,
    Uint8Array.from([0x00]),
  );
  await waitReq1(mock);
  await session.call("g_handshake_interrupt_handler", {
    registers: { ...BASE_REGS },
  });
  return io;
}

test("13h は指定バイトをビッグエンディアンで返す", async () => {
  await withCase(async (s, mock) => {
    s.writeWord(WORD_ADDR, 0x1234);
    s.writeWord(WORD_ADDR + 1, 0xabcd);
    const data = [0x12, 0x34, 0xab, 0xcd];
    const reply = await callRead(mock, BYTE_ADDR, 4);
    expect(Array.from(reply)).toEqual(data);
    s.expectRegisters({ R0: 0, R4: 0x4444 });
  });
});

test("13h はワード 8000h（バイト 10000h、byte_hi の LSB）を読む", async () => {
  await withCase(async (s, mock) => {
    const word = 0x8000;
    s.writeWord(word, 0xa5a5);
    const reply = await callRead(mock, word * 2, 2);
    expect(Array.from(reply)).toEqual([0xa5, 0xa5]);
    s.expectRegisters({ R0: 0, R4: 0x4444 });
  });
});

test("13h バイト数 0 はデータなしで完了する", async () => {
  await withCase(async (s, mock) => {
    const reply = await callRead(mock, BYTE_ADDR, 0);
    expect(reply.length).toBe(0);
    s.expectRegisters({ R4: 0x4444 });
  });
});

/**
 * パターンバイト列をユーザ RAM に書く。
 * @param s セッション
 * @param byteAddr 開始バイトアドレス
 * @param n バイト数
 * @returns 書いた列
 */
function fillPattern(
  s: Mn1613AsmSession,
  byteAddr: number,
  n: number,
): number[] {
  const data: number[] = [];
  for (let i = 0; i < n; i += 1) {
    const b = (i * 13 + 7) & 0xff;
    data.push(b);
    const waddr = (byteAddr + i) >>> 1;
    const cur = s.readWord(waddr);
    if ((byteAddr + i) & 1) {
      s.writeWord(waddr, (cur & 0xff00) | b);
    } else {
      s.writeWord(waddr, (b << 8) | (cur & 0x00ff));
    }
  }
  return data;
}

test("13h は 256 バイトを返す", async () => {
  await withCase(async (s, mock) => {
    const n = 256;
    const expected = fillPattern(s, BYTE_ADDR, n);
    const reply = await callRead(mock, BYTE_ADDR, n);
    expect(Array.from(reply)).toEqual(expected);
    s.expectRegisters({ R0: 0, R4: 0x4444 });
  });
});

test("13h は実用サイズ（512B、上限8KB以内）を返す", async () => {
  await withCase(async (s, mock) => {
    if (MEMREAD_TEST_BYTES > MAX_MEMREAD_TEST_BYTES) {
      throw new Error("MEMREAD_TEST_BYTES must be <= 8KB");
    }
    const expected = fillPattern(s, BYTE_ADDR, MEMREAD_TEST_BYTES);
    const reply = await callRead(mock, BYTE_ADDR, MEMREAD_TEST_BYTES);
    expect(reply.length).toBe(MEMREAD_TEST_BYTES);
    expect(Array.from(reply)).toEqual(expected);
    s.expectRegisters({ R0: 0, R4: 0x4444 });
  });
});

File: retrocpu_boot_monitor/test/mn1613/handshake/handshake_read_memory_test.rs
use std::sync::Arc;

use retrocpu_test_framework_rs::{
    create_session_from_settings, CallOptions, CallRegisters, CodeTestIoMockEntry, FrameworkError,
    IoBoardHandshakeMock, JsonTestSettings, Mn1613AsmSession,
};

const WORD_ADDR: u16 = 0x1800;
const BYTE_ADDR: u32 = (WORD_ADDR as u32) * 2;
const MAX_MEMREAD_TEST_BYTES: usize = 0x2000;
const MEMREAD_TEST_BYTES: usize = 0x0200;

fn base_regs() -> CallRegisters {
    CallRegisters {
        r2: Some(0x2222),
        r3: Some(0x3333),
        r4: Some(0x4444),
        ..Default::default()
    }
}

fn handshake_settings() -> JsonTestSettings {
    let mut s = super::mn1613_rs_settings();
    s.io_mock = Some(vec![CodeTestIoMockEntry::Handshake]);
    s.max_cycles = Some(250_000_000);
    s
}

fn with_case<F>(f: F) -> Result<(), FrameworkError>
where
    F: FnOnce(&mut Mn1613AsmSession, Arc<IoBoardHandshakeMock>) -> Result<(), FrameworkError>,
{
    let mut s = create_session_from_settings(&handshake_settings(), None)?;
    s.run_init()?;
    let mock = s.require_handshake_mock()?;
    let ret = f(&mut s, Arc::clone(&mock));
    s.detach_io_mock();
    ret
}

fn mem_read_header(byte_addr: u32, byte_count: usize) -> Vec<u8> {
    vec![
        0x13,
        ((byte_addr >> 24) & 0xff) as u8,
        ((byte_addr >> 16) & 0xff) as u8,
        ((byte_addr >> 8) & 0xff) as u8,
        (byte_addr & 0xff) as u8,
        (((byte_count as u32) >> 24) & 0xff) as u8,
        (((byte_count as u32) >> 16) & 0xff) as u8,
        (((byte_count as u32) >> 8) & 0xff) as u8,
        ((byte_count as u32) & 0xff) as u8,
        0,
    ]
}

fn call_read(
    s: &mut Mn1613AsmSession,
    mock: &Arc<IoBoardHandshakeMock>,
    byte_addr: u32,
    byte_count: usize,
) -> Result<Vec<u8>, FrameworkError> {
    let header = mem_read_header(byte_addr, byte_count);
    mock.run_io_handler_exchange_ext(&header, byte_count, Some(&[0x00]), || {
        let _ = s.call(
            "g_handshake_interrupt_handler",
            CallOptions {
                registers: Some(base_regs()),
                ..Default::default()
            },
        )?;
        Ok(())
    })
}

fn write_byte_at(s: &mut Mn1613AsmSession, byte_addr: u32, b: u8) {
    let waddr = (byte_addr >> 1) as u16;
    let cur = s.read_word(waddr);
    if (byte_addr & 1) != 0 {
        s.write_word(waddr, (cur & 0xff00) | b as u16);
    } else {
        s.write_word(waddr, ((b as u16) << 8) | (cur & 0x00ff));
    }
}

fn fill_pattern(s: &mut Mn1613AsmSession, byte_addr: u32, n: usize) -> Vec<u8> {
    let mut out = Vec::with_capacity(n);
    for i in 0..n {
        let b = ((i * 13 + 7) & 0xff) as u8;
        out.push(b);
        write_byte_at(s, byte_addr + i as u32, b);
    }
    out
}

#[test]
fn read_memory_returns_big_endian_bytes() -> Result<(), FrameworkError> {
    with_case(|s, mock| {
        s.write_word(WORD_ADDR, 0x1234);
        s.write_word(WORD_ADDR + 1, 0xabcd);

        let reply = call_read(s, &mock, BYTE_ADDR, 4)?;
        assert_eq!(reply, vec![0x12, 0x34, 0xab, 0xcd]);
        s.expect_registers(
            &CallRegisters {
                r0: Some(0),
                r4: Some(0x4444),
                ..Default::default()
            },
            None,
        )
    })
}

#[test]
fn read_memory_can_access_word_0x8000() -> Result<(), FrameworkError> {
    with_case(|s, mock| {
        let word = 0x8000u16;
        s.write_word(word, 0xa5a5);

        let reply = call_read(s, &mock, (word as u32) * 2, 2)?;
        assert_eq!(reply, vec![0xa5, 0xa5]);
        s.expect_registers(
            &CallRegisters {
                r0: Some(0),
                r4: Some(0x4444),
                ..Default::default()
            },
            None,
        )
    })
}

#[test]
fn read_memory_zero_count_returns_no_data() -> Result<(), FrameworkError> {
    with_case(|s, mock| {
        let reply = call_read(s, &mock, BYTE_ADDR, 0)?;
        assert!(reply.is_empty());
        s.expect_registers(
            &CallRegisters {
                r4: Some(0x4444),
                ..Default::default()
            },
            None,
        )
    })
}

#[test]
fn read_memory_returns_256_bytes() -> Result<(), FrameworkError> {
    with_case(|s, mock| {
        let n = 256usize;
        let expected = fill_pattern(s, BYTE_ADDR, n);
        let reply = call_read(s, &mock, BYTE_ADDR, n)?;
        assert_eq!(reply, expected);
        s.expect_registers(
            &CallRegisters {
                r0: Some(0),
                r4: Some(0x4444),
                ..Default::default()
            },
            None,
        )
    })
}

#[test]
fn read_memory_returns_512_bytes() -> Result<(), FrameworkError> {
    with_case(|s, mock| {
        assert!(MEMREAD_TEST_BYTES <= MAX_MEMREAD_TEST_BYTES);
        let expected = fill_pattern(s, BYTE_ADDR, MEMREAD_TEST_BYTES);
        let reply = call_read(s, &mock, BYTE_ADDR, MEMREAD_TEST_BYTES)?;
        assert_eq!(reply.len(), MEMREAD_TEST_BYTES);
        assert_eq!(reply, expected);
        s.expect_registers(
            &CallRegisters {
                r0: Some(0),
                r4: Some(0x4444),
                ..Default::default()
            },
            None,
        )
    })
}
